### DISCUSSION TITLE
dual ethernet wiznet5500 esp32s3

### DISCUSSION THREAD

**Discussion by littlegreenbean33** [micropython/micropython#16491] [ESP32] [answered]
Hi,

I am trying to set up an esp32S3 with dual ethernet interfaces. My use case requires bridging at OSI7 level 2 different networks.

In the discord  channel I have been sharing my experience ( and received help).
https://discord.com/channels/574275045187125269/1012726673709469818/1322641406573019269

I know IDF support more than one ethernet interface. I was just wondering if that functionality has been relayed into micropython ?

Any ideas ?

Currently both interfaces show different MAC addresses . However, same IP  and IP connection details.....even when connected to 2 different routers with different ip ranges.

Regards

JC
(EDITED. system return different MACs still same IP configuration, connected to 2 separate  routers)

**Comment by littlegreenbean33** [comment]
my notes above maybe incorrect. It looks like only one LAN interface is possible ?

**Comment by robert-hh** [comment]
Does it make a difference if you set the IP a dresses manually with ifconfig()?

**Reply by littlegreenbean33** [reply]
will try but

```
>>> lan1 is lan2 
True
```

spells for the wrong result.

**Reply by littlegreenbean33** [reply]
do you think I should open an bug report ?

I should be able to have multiple interfaces on the same medium. i.e. PPP I could have 3 IP over serial connections....

I can only understand may be related to this... ignorance is a bliss...

```
135     if (args[ARG_id].u_obj != mp_const_none) {
136         if (mp_obj_get_int(args[ARG_id].u_obj) != 0) {
137             mp_raise_ValueError(MP_ERROR_TEXT("invalid LAN interface identifier"));
138         }
139     }
```
in network_lan.c is supposed to rise an error if id != 0 ?? but that is not happening to me. I use 1 and 0.

**Reply by robert-hh** [reply, marked as answer]
Looking into the code I see, that the ESP32 port uses a static object for LAN. So yes, one can only have a single LAN interface. It may be possible to use a dynamic object instead, but for that the code must be changed (not a big deal) and tested. The tiny downside of that is, that dynamic objects get lost on soft reset.

**Reply by littlegreenbean33** [reply]
thanks @robert-hh. Can you point me to where in the code this is ?  ...My C/Unix system programming is 25 years old...but I may entertain myself looking to see if there is anything I can do....if only testing in case somebody else takes the lead.

I'll open a feature request though.

Thanks for looking into this.

JC

**Reply by robert-hh** [reply]
The file is ports/esp32/network.lan.c. The LAN object is defined at line 48-64. It is instantiated at line 107, where the dynamic object would have to be assigned. Using something like:
`self = mp_obj_malloc(lan_if_obj_t, &lan_if_type);` and then setting it up accordingly.
 But there is more to it, and especially the espressif stack must be able to handle more interfaces.

**Reply by littlegreenbean33** [reply]
@robert-hh  idf does support multiple LAN. I checked  before my tests.  There is a patch late 2022, 2023 with the changes. I will look for it and share the link

**Reply by littlegreenbean33** [reply]
I decided to tes the dual port based on these comments

https://github.com/espressif/esp-idf/issues/7318

https://github.com/espressif/esp-idf/commit/ef3038490275d02dee46319f9c596d807163213f
